File: tests/cpp/test_eigenCG.cpp
/******************************************************************************/
/*                                                                            */
/*                            gstlearn C++ Library                            */
/*                                                                            */
/* Copyright (c) (2023) MINES Paris / ARMINES                                 */
/* Authors: gstlearn Team                                                     */
/* Website: https://gstlearn.org                                              */
/* License: BSD 3-clause                                                      */
/*                                                                            */
/******************************************************************************/
#include "Basic/File.hpp"
#include "LinearOp/ScaleOp.hpp"
#include "LinearOp/LinearOpCGSolver.hpp"

#include <iostream>
#include <Eigen/IterativeLinearSolvers>
#include <unsupported/Eigen/IterativeSolvers>

int main(int argc, char *argv[])
{
  std::stringstream sfn;
  sfn << gslBaseName(__FILE__) << ".out";
  StdoutRedirect sr(sfn.str(), argc, argv);

  int n = 5;

  VectorDouble b({2, 4, 6, 8, 10});

  VectorDouble o(n);

  VectorDouble c(n);

  ScaleOp I(n, 2.0);
  LinearOpCGSolver<ScaleOp> s(&I);

  std::cout << "b = " << b.toString();
  s.solve(b, o);
  std::cout << "o = " << o.toString();

  I.evalDirect(o, c);
  std::cout << "c = " << c.toString();

}

File: tests/py/test_eigenCG.py
import gstlearn as gl

n = 5

b = [2, 4, 6, 8, 10]
B = gl.VectorEigen(b)

o = gl.VectorDouble(n, 0)
O = gl.VectorEigen(n)

c = gl.VectorDouble(n, 0)
C = gl.VectorEigen(n)

I = gl.ScaleOp(n, 2.)
s = gl.LinearOpCGSolver(I)

print(b)

s.solve(b, o)
print(o)

s.solve(B, O)
print(O.getValues())

I.evalDirect(o, c)
print(c)

I.evalDirect(O, C)
print(C.getValues())
